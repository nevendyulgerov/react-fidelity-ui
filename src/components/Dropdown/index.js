import Dropdown from './script/Dropdown';
import { changeSingleSelect, changeMultiSelect, deselectItem, filterItemsByName, syncDropdownMenuOffset } from './script/Utils';

export { changeSingleSelect, changeMultiSelect, deselectItem, filterItemsByName, syncDropdownMenuOffset };
export default Dropdown;
