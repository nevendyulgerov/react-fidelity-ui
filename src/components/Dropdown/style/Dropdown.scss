
/**
 * Component: Dropdown
 */

[data-component="dropdown"] {
  position: relative;
  margin: 0 34px 0 0;

  &.disabled {
    .trigger.toggle-menu {
      pointer-events: none;
      opacity: 0.5;
      transition: all 0.15s ease-in-out;
    }
  }

  .trigger.toggle-menu {
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;
    outline: none;
    background-color: $white;
    color: $grey;
    border-radius: 4px;
    padding: 12px 16px;
    font-weight: 600;
    font-size: 14px;
    cursor: pointer;
    border: 1px solid transparent;
    box-shadow: $shadow-depth-0;
    transition: all 0.15s ease-in-out;

    &.active {
      box-shadow: $shadow-depth-1;
      border: 1px dashed $grey;
      transition: all 0.15s ease-in-out;

      .icon-box {
        [data-component="icon"] {
          transform: rotate(180deg);
          transition: all 0.15s ease-in-out;
        }
      }
    }

    &:hover {
      box-shadow: $shadow-depth-1;
      transition: all 0.15s ease-in-out;
    }

    .icon-box {
      position: relative;

      [data-component="icon"] {
        opacity: 0;
        pointer-events: none;
        position: relative;
        left: 8px;
        transition: all 0.15s ease-in-out;

        &.active {
          opacity: 1;
          transition: all 0.15s ease-in-out;
        }
      }
    }

    [data-component="loader"] {
      position: absolute;
      top: -1px;
      left: 7px;
    }

    div.text {
      width: calc(100% - 24px);
      text-align: left;
    }

    span.text {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
  }

  .menu {
    opacity: 0;
    pointer-events: none;
    min-width: calc(100% + 34px);
    width: auto;
    position: absolute;
    left: 0;
    top: 70px;
    background-color: $white;
    box-shadow: $shadow-depth-2;
    border-radius: 4px;
    z-index: 999;
    transition: all 0.15s ease-in-out;

    &.active {
      opacity: 1;
      pointer-events: all;
      transition: all 0.15s ease-in-out;
    }

    .menu-heading {
      padding: 10px 15px;
      border-bottom: 1px solid $silver;

      .component-title {
        display: block;
        font-weight: 800;
      }

      .component-text {
        display: block;
        font-size: 14px;
        font-weight: 300;
      }
    }

    .menu-body {
      border-bottom: 1px solid $silver;

      &.no-results {
        border: none;

        .filter {
          box-shadow: none;
        }
      }

      .filter {
        position: relative;
        width: calc(100% - 30px);
        min-width: 200px;
        padding: 10px 15px 12px 15px;
        border-bottom: 1px solid $silver;
        box-shadow: 0 0 1px 0 rgba(0,0,0,0.16),0 0 0 0 rgba(239,239,239,0.12);

        input {
          width: calc(100% - 16px);
          border-radius: 4px;
          border: 0;
          padding: 12px 8px 12px 8px !important;
          box-shadow: $shadow-depth-0;
          background-color: $soft-grey;
          font-size: 14px;
          transition: all 0.15s ease-in-out;

          &.addable {
            padding: 12px 60px 12px 8px !important;
            width: calc(100% - 68px);
          }

          &.active {
            display: block;
          }

          &:focus {
            border: 0;
            outline: 0;
            box-shadow: $shadow-depth-1;
            transition: all 0.15s ease-in-out;
          }

          &:active {
            border: 0;
            outline: 0;
          }
        }

        .trigger.add-item {
          opacity: 0;
          pointer-events: none;
          position: absolute;
          top: 16px;
          right: 22px;
          border: 0;
          border-radius: 4px;
          background-color: $soft-green;
          color: $white;
          padding: 6px 10px;
          box-shadow: $shadow-depth-0;
          transition: all 0.15s ease-in-out;

          &.active {
            opacity: 1;
            pointer-events: all;
            transition: all 0.15s ease-in-out;
          }

          &:hover {
            box-shadow: $shadow-depth-1;
            transition: all 0.15s ease-in-out;
          }
        }
      }

      ul {
        max-height: 300px;
        overflow-x: hidden;
        overflow-y: auto;
        padding: 10px 15px 10px 15px;
        margin: 0;
        list-style-type: none;

        &.no-items {
          padding: 0;
        }

        li {
          display: block;
          padding: 0;
          margin: 0 0 10px 0;

          &::selection {
            background-color: transparent;
          }

          &:last-child {
            margin: 0;
          }

          .checkbox {
            display: flex;
            justify-content: flex-start;
            align-items: center;
            margin: 0;

            label {
              font-weight: 800;
              font-size: 14px;
              white-space: nowrap;
              overflow: hidden;
              text-overflow: ellipsis;

              &::selection {
                background-color: transparent;
              }

              &:before {
                top: 4px;
              }

              &:after {
                top: 8px;
              }
            }
          }
        }
      }
    }
  }

  .inline-tags {
    display: flex;
    justify-content: flex-start;
    flex-wrap: wrap;

    &.has-tags {
      position: relative;
      top: 4px;
    }
  }
}
