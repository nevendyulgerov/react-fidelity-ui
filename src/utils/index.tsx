/**
 * @description Is null
 * @param val
 */
export const isNull = (val: any) : boolean => val === null;

/**
 * @description Is array
 * @param val
 */
export const isArr = (val: any) : boolean => Array.isArray(val);

/**
 * @description Is object
 * @param val
 */
export const isObj = (val: any) : boolean => typeof val === 'object' && !isArr(val) && !isNull(val);

/**
 * @description Is number
 * @param val
 */
// eslint-disable-next-line no-restricted-globals
export const isNum = (val: any) : boolean => typeof val === 'number' && !isNaN(val);

/**
 * @description Is function
 * @param val
 */
export const isFunc = (val: any) : boolean => typeof val === 'function';

/**
 * @description Is string
 * @param val
 */
export const isStr = (val: any) : boolean => typeof val === 'string';

/**
 * @description Is undefined
 * @param val
 */
export const isUndef = (val: any) : boolean => typeof val === 'undefined';

/**
 * @description Is boolean
 * @param val
 */
export const isBool = (val: any) : boolean => typeof val === 'boolean';

/**
 * @description Is non empty string
 * @param val
 */
export const isNonEmptyStr = (val: any) : boolean => isStr(val) && val !== '';

/**
 * @description Uid
 * @param len
 */
export const uid = (len: number = 7): string => Math.random().toString(35).substr(2, len);

/**
 * @description Get object keys
 * @param obj
 * @returns {Array}
 */
export const getKeys = (obj: any) => Object.keys(obj);

/**
 * @description Each key
 * @param obj
 * @param callback
 */
export const eachKey = (obj: any, callback: (key: string, prop: any, index: number) => void) => {
  getKeys(obj).forEach((key, index) => callback(key, obj[key], index));
};

/**
 * @description Observe intersection
 * @param target
 * @param callback
 */
export const observeIntersection = (target: HTMLElement, callback: (items: []) => void) => {
  if (!isFunc(window.IntersectionObserver)) {
    return console.warn('IntersectionObserver is not supported.');
  }

  // eslint-disable-next-line compat/compat
  const Observe = new IntersectionObserver((items: []) => {
    callback(items);
  });

  Observe.observe(target);

  return Observe;
};

/**
 * @description Get css var
 * @param varName
 */
export const getCssVar = (varName: string): string | null => (
  getComputedStyle(document.documentElement).getPropertyValue(varName)
);

/**
 * @description Set css var
 * @param varName
 * @param varValue
 */
export const setCssVar = (varName: string, varValue: string | null) => {
  document.documentElement.style.setProperty(varName, varValue);
};

/**
 * @description Limit number
 * @param num
 * @param max
 */
export const limitNum = (num: number, max: number) => (num <= max ? num : max);

/**
 * @description Can render prop
 * @param prop
 */
export const isRenderable = (prop: any) => isNonEmptyStr(prop) || isNum(prop) || isObj(prop);
