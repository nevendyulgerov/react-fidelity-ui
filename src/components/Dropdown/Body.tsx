import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { isNonEmptyStr } from '../../utils';

export type DropdownBodyProps = {
  children: React.ReactNode,
  size?: string | 'auto',
  className?: string | null,
  active?: boolean | false,
  [key: string]: any
};

const sizes = ['auto', 'xs', 'sm', 'md', 'lg', 'xl'];

const DropdownBody = (props: DropdownBodyProps) => {
  const { children, size, className, active, ...restProps } = props;
  const componentClassName: string = classNames({
    dropdown__body: true,
    'dropdown__body--active': active,
    // @ts-ignore
    [`dropdown__body--size-${size}`]: DropdownBody.sizes.includes(size),
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  return (
    <div
      aria-expanded={active}
      {...restProps}
      className={componentClassName}
    >
      {children}
    </div>
  );
};

DropdownBody.sizes = sizes;

DropdownBody.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  size: PropTypes.oneOf(sizes),
  className: PropTypes.string,
  active: PropTypes.bool
};

DropdownBody.defaultProps = {
  size: 'auto',
  className: null,
  active: false
};

export default DropdownBody;
