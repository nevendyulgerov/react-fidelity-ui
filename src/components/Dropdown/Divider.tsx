import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { isNonEmptyStr } from '../../utils';

export type DropdownDividerProps = {
  className?: string | null,
  [key: string]: any
}

const DropdownDivider = ({ className, ...restProps }: DropdownDividerProps) => {
  const componentClassName: string = classNames({
    dropdown__divider: true,
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  return (
    <div
      {...restProps}
      className={componentClassName}
    />
  );
};

DropdownDivider.propTypes = {
  className: PropTypes.string
};

DropdownDivider.defaultProps = {
  className: null
};

export default DropdownDivider;
