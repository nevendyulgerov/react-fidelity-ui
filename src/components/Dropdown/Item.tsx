import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { isNonEmptyStr } from '../../utils';

export type DropdownItemProps = {
  className?: string | null,
  active?: boolean,
  renderItem: (className: string) => React.ReactNode
};

const DropdownItem = (props: DropdownItemProps) => {
  const { className, active, renderItem } = props;

  const componentClassName = classNames({
    dropdown__item: true,
    'dropdown__item--active': active,
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  return renderItem(componentClassName);
};

DropdownItem.propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  renderItem: PropTypes.func.isRequired
};

DropdownItem.defaultProps = {
  className: null,
  active: false,
  renderItem: () => {}
};

export default DropdownItem;
