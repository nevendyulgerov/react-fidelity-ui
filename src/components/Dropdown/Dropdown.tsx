import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import Body from './Body';
import Text from './Text';
import Item from './Item';
import Divider from './Divider';
import { isNonEmptyStr, isFunc, isObj } from '../../utils';

const { addEventListener, removeEventListener } = window;
const { useRef, useEffect } = React;

export type DropdownProps = {
  children: React.ReactNode,
  className?: string | null,
  active?: boolean | false,
  onClickOutside?: () => void | null,
  [key: string]: any
};

const Dropdown = (props: DropdownProps) => {
  const { children, className, active, onClickOutside, ...restProps } = props;
  const refComponent = useRef(null);
  const latestOnClickOutside = useRef(onClickOutside);
  const hasOnClickOutside: boolean = isFunc(onClickOutside);
  const componentClassName: string = classNames({
    dropdown: true,
    'dropdown--active': active,
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  const onClick = (event: MouseEvent): void => {
    const dropdownNode = refComponent.current;
    // @ts-ignore
    const closestDropdown = event.target.closest('.dropdown');

    if (!isObj(closestDropdown) || !closestDropdown.isSameNode(dropdownNode)) {
      // @ts-ignore
      onClickOutside();
    }
  };

  useEffect(() => {
    latestOnClickOutside.current = onClickOutside;

    if (hasOnClickOutside) {
      addEventListener('click', onClick);
    }

    return () => {
      if (hasOnClickOutside) {
        removeEventListener('click', onClick);
      }
    };
  }, [onClickOutside]);

  return (
    <div
      {...restProps}
      ref={refComponent}
      className={componentClassName}
    >
      {children}
    </div>
  );
};

Dropdown.Body = Body;
Dropdown.Text = Text;
Dropdown.Item = Item;
Dropdown.Divider = Divider;

Dropdown.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  className: PropTypes.string,
  active: PropTypes.bool,
  onClickOutside: PropTypes.func
};

Dropdown.defaultProps = {
  className: null,
  active: false,
  onClickOutside: null
};

export default Dropdown;
