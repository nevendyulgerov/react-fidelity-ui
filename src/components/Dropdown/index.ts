import Divider from './Divider';
import Dropdown from './Dropdown';
import Item from './Item';
import Text from './Text';
import Title from './Title';

export {
  Divider,
  Dropdown,
  Item,
  Text,
  Title
};
