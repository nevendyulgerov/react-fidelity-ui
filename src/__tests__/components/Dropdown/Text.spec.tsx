import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Dropdown from '../../../components/Dropdown';
import { isObj } from '../../../utils';

describe('Components / Dropdown / Text', () => {
  it('matches snapshot', () => {
    const { baseElement } = render(
      <Dropdown.Text>
        Dropdown Text
      </Dropdown.Text>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('displays children', () => {
    const { baseElement } = render(
      <Dropdown.Text>
        <h1>Heading 1</h1>
        <h2>Heading 2</h2>
      </Dropdown.Text>
    );
    const node = baseElement.querySelector('.dropdown__text');

    // @ts-ignore
    expect(node.children.length).toBe(2);
  });

  it('attaches given className', () => {
    const className = 'dropdown__text-xyz';
    const { baseElement } = render(
      <Dropdown.Text className={className}>
        Dropdown
      </Dropdown.Text>
    );
    const node = baseElement.querySelector(`.${className}`);

    expect(isObj(node)).toBe(true);
  });

  it('adds given style prop', () => {
    const style = { height: '20px' };
    const { baseElement } = render(
      <Dropdown.Text style={style}>
        Dropdown
      </Dropdown.Text>
    );
    const node = baseElement.querySelector('.dropdown__text');

    // @ts-ignore
    expect(getComputedStyle(node).height).toBe(style.height);
  });

  it('adds given callback prop', () => {
    let isClickTriggered = false;
    const onClick = () => {
      isClickTriggered = true;
    };
    const { baseElement } = render(
      <Dropdown.Text onClick={onClick}>
        Dropdown
      </Dropdown.Text>
    );
    const node = baseElement.querySelector('.dropdown__text');

    // @ts-ignore
    fireEvent.click(node);

    expect(isClickTriggered).toBe(true);
  });

  it('adds given role prop', () => {
    const role = 'dropdown-text';
    const { baseElement } = render(
      <Dropdown.Text role={role}>
        Dropdown
      </Dropdown.Text>
    );
    const node = baseElement.querySelector('.dropdown__text');

    // @ts-ignore
    expect(node.getAttribute('role')).toBe(role);
  });
});
