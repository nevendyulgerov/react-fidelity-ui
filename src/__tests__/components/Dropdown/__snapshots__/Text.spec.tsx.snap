// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Dropdown / Text matches snapshot 1`] = `
<body>
  <div>
    <div
      class="dropdown__text"
    >
      Dropdown Text
    </div>
  </div>
</body>
`;
